### PR TITLE
Docs restructure (top level plugin sections)

### PR DESCRIPTION
**Pull request by eysi09**
## What this PR does / why we need it

This PR restructures our documentation to make Garden easier to adopt. 

## Special notes for your reviewer

The goal is to make Garden easier to adopt via better documentation. Below are some notes on how we achieve that goal. 

> Note that this is mostly based on existing docs. If you find things that are missing, they were most likely missing before. This new structure only makes it more visible (see note on visibility below). The scope here is strictly to fix the structure, not re-write everything along the way, although I did my best to fill in the most gaping blanks as well as align text to the new structure. 

### Prefer step-by-step instructions

The current version of our documentation contains a lot of long format text with no clear step-by-step instructions. The current In-Cluster Building and Remote Kubernetes guides are good (or bad?) examples of this. 

With this new structure, I've tried to list requirements up front, then include a step-by-step guide for meeting these requirements, and then finally how to feed them back to Garden.

The [Remote K8s Plugin Configuration guide](https://docs.garden.io/v/docs-restructure-alt/kubernetes-plugins/remote-k8s) is a good example of this. 

### Plugins first

The new structure makes plugins a "first class citizen". Previously, plugin concepts were bleeding into Garden concepts and vice versa. Here I attempt to completely separate the two. This should make it easier for the user to understand the border between Garden and plugins and to more easily find what they're looking for.

If you're using the K8s plugin, that's where you look for K8s topics. As opposed to in a general purpose "Guides" section that mixes everything together. 

Each plugin section also follow the same format, making it easier to navigate the docs and adopt new plugins.

### Consistency

The idea behind the re-structure is that you adopt Garden like this (simplified):

- You start by picking your plugin(s).
- Once you pick a plugin, you check out its requirements, follow step-by-step guides for meeting those requirements, then configure the plugin.
- Finally you pick your module types and then configure them.

This is why all the plugin sections have the same consistent format:

- First there's a "How it Works" guide
- Then "Plugin Configuration" guide(s)
- Then "Module Configuration" guide(s)
- Then a section for advanced guides
 
For simple plugins we merge this into a single page but ideally it's split up in general so it's easy to find what you're looking for by just looking at the section names. 

This should be consistent throughout, and once you've added your first plugin, you basically know how to add the rest. 

I'd go so far as to argue that if a given Garden functionality doesn't map neatly to a consistent doc structure, then that suggests the functionality is too complex and needs more thought. 

### Visibility, atomicity and scalability

This new structure makes it visible where docs are not good enough or simply missing. Take e.g. the guide for creating a cluster on [Azure](https://docs.garden.io/v/docs-restructure-alt/kubernetes-plugins/remote-k8s/create-cluster/azure).

This is intentional and I'd rather want it to sting until we fill in the missing guides as opposed to having a structure that obfuscates the fact that a guide is missing in the first place. Note that both restructuring docs _and_ filling in missing docs is out of scope for this one.

Another benefit is that it's much easier to point users to specific content. Often you're not starting completely from scratch, you may have a cluster but just need to configure that GCR registry. The new structure should make it easy to find exactly that guide. 

Finally, this is hopefully more scalable and will make it a lot clearer were to add new content. I e.g. think that we can make more "surgical" updates for the SGv2 docs than we could've before. 

## Open Questions

- The pages for configuring providers are called "Plugin Configuration". The plan is to re-name `providers` to `plugins` and I felt that this made how Garden works a bit clearer, since there's a lot of reference to plugins in general. But we can also call them "Provider Configuration" if this feels confusing. Thoughts?
- I made the main plugins their own top-level section and then added a section for "Other Plugins". There's an alternate [version here](https://docs.garden.io/v/docs-restructure/) were "Plugins" is the top level section and all plugins nested under there. I opted away from that because I felt there was too much nesting. The former scales a bit better though but I personally think that's not a concern for now and would rather reduce nesting. Does that make sense?

## Must have To Dos before merging

- [x] Update redirects
- [x] Make sure there are no broken links
- [x] Update commit message
- [x] Fix typos! (I'll try and go through this and fix what I can to spare my trusted reviewers the effort.)
- ~Add a separate GitBook "space" for the previous version so users can fallback to that if they're not happy with this one. (Can technically be done after we merge.)~ EDIT: On second thought we may not need this. But we can easily spin up if requested.

## Future work

There's obviously a lot of future work to be done but highlighting things that could be in this PR. But I just got to draw the line somewhere, this has already taken awhile and I'd rather merge and iterate.

- Better notes on how to read the docs in the "Welcome" guide
- Re-name "In-Cluster Building" guide to "Build Modes" since it also contains a guide on local docker. Or otherwise split up. 
- Pin an issue with an initial list of known missing docs and then solicit more items from community. We can then tick them off one-by-one. 
- The docs for deploying the `container` module are under Kubernetes Plugins > Module Configuration > Container. The previous [catch-all docs](https://docs.garden.io/guides/container-modules) we had on the container module are otherwise mostly under Plugins/Container. This is a little awkward and we can probably do better. 
- I did not update the "Using Garden" sections at all. They need to be revised, but maybe that's better left until we have SGv2.

### CODE REVIEW

**Comment by eysi09** [comment]
Note that this is the alternate version to #3373 which does not have top level plugin sections. If we prefer #3373 we can close this and use that one instead.

**Comment by eysi09** [comment]
Oh, and here's a link to this version: https://docs.garden.io/v/docs-restructure-alt/

**Comment by thsig** [comment]
> The pages for configuring providers are called "Plugin Configuration". The plan is to re-name providers to plugins and I felt that this made how Garden works a bit clearer, since there's a lot of reference to plugins in general. But we can also call them "Provider Configuration" if this feels confusing. Thoughts?

To clarify: A plugin may introduce a provider configuration, and one or more module/action types.

I.e. there's no such thing as a plugin configuration: Plugins expose provider, module & action configurations.

That said, calling the pages "Plugin Configuration" etc. could still make sense, if the meaning is "Plugin-specific Configuration = Provider + Module/Action Configuration".

**Comment by eysi09** [comment]
> I.e. there's no such thing as a plugin configuration: Plugins expose provider, module & action configurations. Users configure the providers, modules and actions introduced by plugins.

Except that we're planning to rename providers to plugins. 

I guess my thinking was that we'd just stick with calling it "plugin configuration" as opposed to "provider configuration" and avoid having to explain the difference. So when you glance the docs and see something like "Kubernetes plugin configuration" it's obvious that that's your starting point whereas "provider configuration" feels more opaque.

So in a sense I'm omitting some detail so that it feels more intuitive to a new user. 

Feels like: "Pick a plugin, configure it, and add modules" is a simpler and more familiar mental model than "Pick a plugin, configure the provider exposed by the plugin, and add modules". 

All that being said, we can stick with "configure providers" until we actually make this change if the above doesn't make sense.

**Comment by edvald** [comment]
Just to chime in re: plugin vs. provider configuration, I agree with @thsig, we should stick with "Provider configuration" for the time being, until we make the planned changes in code. But I also agree that the extra terminology feels confusing and that we _should_ get to changing it sooner than later.

File: docs/k8s-plugins/advanced/in-cluster-building.md [hunk 1]
```diff
@@ -244,14 +246,57 @@ clusterDocker:
...
 
 Make sure your cluster has enough resources and storage to support the required services, and keep in mind that these
-services are shared across all users of the cluster. Please look at the [resources](../reference/providers/kubernetes.md#providersresources) and [storage](../reference/providers/kubernetes.md#providersstorage) sections in the provider reference for
+services are shared across all users of the cluster. Please look at the [resources](../../../reference/providers/kubernetes.md#providersresources) and [storage](../../../reference/providers/kubernetes.md#providersstorage) sections in the provider reference for
 details.
 
 ### Local Docker
 
-This is the default mode. It is the least efficient one for remote clusters, but requires no additional configuration or services to be deployed to the cluster. For remote clusters, you do however need to explicitly configure a [deployment registry](#configuring-a-deployment-registry), and obviously you'll need to have Docker running locally.
+This is the default build mode. It is the least efficient one for remote clusters, but requires no additional services
```
Comment (thsig) [inline docs/k8s-plugins/advanced/in-cluster-building.md:254]:
```suggestion
This is the default build mode. It is usually the least efficient one for remote clusters, but requires no additional services
```

**Comment by thsig** [comment]
> I guess my thinking was that we'd just start calling it "plugin configuration" right away, as opposed to "provider configuration", and avoid having to explain the relationship between plugins and providers.

Yeah, makes sense. This looks good to me!

File: docs/README.md [hunk 2]
```diff
@@ -28,34 +27,71 @@
...
+## ☘️ Pulumi Plugin
+
+* [About](./pulumi-plugin/about.md)
+* [Plugin Configuration](./pulumi-plugin/configure-plugin.md)
+* [Module Configuration](./pulumi-plugin/configure-modules.md)
+
+## 🌹 Other Plugins
+
+* [Container](./other-plugins/container.md)
+* [Exec](./other-plugins/exec.md)
```
[the remark this answers is not in the document]
Comment (eysi09) [reply docs/README.md:75]: That's a great point. 

There's no "recommended" grouping though. It's just that plugins with multiple pages have their own grouping whereas the exec one is just a single page. (Plus we're running out of flower emojis.) 

Do you think we should split it into multiple pages and have a specific Exec section? Or have a specific Exec section but only that one page?

Note also that the [other version](https://docs.garden.io/v/docs-restructure/) I mention in the "Open Questions" above has all plugins under a "Plugins" section. So they're all equally prominent.

File: docs/guides/adopting-garden.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  - name: dev
+
+providers:
+ - name: kubernetes
+   environments: ["dev"]
+```
+
+Each plugin has a dedicated section in our documentation that explains how it works and how you can get started using it.
+
+We recommend starting simple with a single plugin and adding more as needed. In most cases you'll want to start with either the `kubernetes` or `local-kubernetes` plugins and build from there.
```
[the remark this answers is not in the document]
Comment (eysi09) [reply docs/guides/adopting-garden.md:31]: That's an interesting point but there's a lot of content already that talks about plugins.  

The GitHub README talks about them and so does the "How Garden works" guide. Same for our website.
 
So it kind of feels like we need to stick with plugins and explain what providers are when you configure them. 

That being said: You could definitely argue that we should've just dropped the plugin notion to being with—until we have a plugin SDK. Make it feel more like a feature of Garden, which it absolutely is. It just feels like that's a bigger change and a refactor in and of itself. 

How about we "take that offline" and address in a separate PR if we decide to take that route?

[the remark this answers is not in the document]
Comment (eysi09) [reply docs/guides/adopting-garden.md:31]: On second thought, maybe it's not as big a change as I think. I'll scan through the other guides and see if your suggestion makes sense.

File: docs/k8s-plugins/advanced/in-cluster-building.md [hunk 4]
```diff
@@ -234,7 +236,7 @@ Enable this mode by setting `buildMode: cluster-docker` in your `kubernetes` pro
...
 
-By default, Garden will install an NFS volume provisioner into `garden-system` in order to be able to efficiently synchronize build sources to the cluster and then attaching those to the Kaniko pods. You can also [specify a storageClass](../reference/providers/kubernetes.md#providersstoragesyncstorageclass) to provide another _ReadWriteMany_ capable storage class to use instead of NFS. This may be advisable if your cloud provider provides a good alternative, or if you already have such a provisioner installed.
+By default, Garden will install an NFS volume provisioner into `garden-system` in order to be able to efficiently synchronize build sources to the cluster and then attaching those to the Kaniko pods. You can also [specify a storageClass](../../../reference/providers/kubernetes.md#providersstoragesyncstorageclass) to provide another _ReadWriteMany_ capable storage class to use instead of NFS. This may be advisable if your cloud provider provides a good alternative, or if you already have such a provisioner installed.
```
[the remark this answers is not in the document]
Comment (eysi09) [reply docs/k8s-plugins/advanced/in-cluster-building.md]: I just removed this block. Also this build mode is deprecated so we'll be removing the whole thing soon. So didn't bother with further adjustments.

File: docs/guides/adopting-garden.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (eysi09) [reply docs/guides/adopting-garden.md:31]: I've updated the page names and some of the text.

**Review by thsig — APPROVED** [review]

**Pull request merged**